internal/internal_eager_activity.go

```diff
@@ -0,0 +1,157 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"fmt"
+	"sync"
+
+	"go.temporal.io/api/workflowservice/v1"
+)
+
+// eagerActivityExecutor is a worker-scoped executor for eager activities that
+// are returned from workflow task completion responses.
+type eagerActivityExecutor struct {
+	eagerActivityExecutorOptions
+
+	activityWorker *activityWorker
+	pendingCount   int // Only access under countLock
+	executingCount int // Only access under countLock
+	countLock      sync.Mutex
+}
+
+type eagerActivityExecutorOptions struct {
+	disabled  bool
+	taskQueue string
+	// If 0, there is no maximum
+	maxConcurrent int
+}
+
+// newEagerActivityExecutor creates a new worker-scoped executor without an
+// activityWorker set. The activityWorker must be set on the responding executor
+// before it will be able to execute activities.
+func newEagerActivityExecutor(options eagerActivityExecutorOptions) *eagerActivityExecutor {
+	return &eagerActivityExecutor{eagerActivityExecutorOptions: options}
+}
+
+func (e *eagerActivityExecutor) applyToRequest(
+	req *workflowservice.RespondWorkflowTaskCompletedRequest,
+) (amountReserved int) {
+	// Go over every command checking for activities that can be eagerly executed
+	for _, command := range req.Commands {
+		if attrs := command.GetScheduleActivityTaskCommandAttributes(); attrs != nil {
+			// If not present, disabled, or on a different task queue, we must mark as
+			// explicitly disabled
+			if e == nil || e.disabled || e.activityWorker == nil || e.taskQueue != attrs.TaskQueue.GetName() {
+				attrs.RequestEagerExecution = false
+			} else if attrs.RequestEagerExecution {
+				// If it has been requested, attempt to reserve one pending
+				attrs.RequestEagerExecution = e.reserveOnePendingSlot()
+				if attrs.RequestEagerExecution {
+					amountReserved++
+				}
+			}
+		}
+	}
+	return
+}
+
+func (e *eagerActivityExecutor) reserveOnePendingSlot() bool {
+	// Lock during count checks. Nothing in here blocks including the channel
+	// receive to serve a slot.
+	e.countLock.Lock()
+	defer e.countLock.Unlock()
+	// Confirm that, if we have a max, pending + executing isn't already there
+	if e.maxConcurrent > 0 && e.executingCount+e.pendingCount >= e.maxConcurrent {
+		// No more room
+		return false
+	}
+	// Reserve a spot for our request via a non-blocking attempt to take a poller
+	// request entry which essentially reserves a spot
+	select {
+	case <-e.activityWorker.worker.pollerRequestCh:
+	default:
+		return false
+	}
+
+	// We can request, so increase the pending count
+	e.pendingCount++
+	return true
+}
+
+func (e *eagerActivityExecutor) handleResponse(
+	resp *workflowservice.RespondWorkflowTaskCompletedResponse,
+	amountReserved int,
```

Comment:
nit: I'd call this `amountActivitySlotsReserved` to clarify

Reply:
Unnecessary IMO since this is already in an eager activity context, but I changed anyways